tactics/tactics.ml

```diff
@@ -315,6 +315,36 @@ let convert_concl ~cast ~check ty k =
     end
   end
 
+let with_reverse_casts env env' sigma c =
+  let ev, l = EConstr.destEvar sigma c in
+  let () = match l with
+    | Nil | Default (_, Nil) -> ()
+    | _ -> assert false
+  in
+  let rec remake_instance acc k = function
+    | [] -> acc
+    | false :: rest -> remake_instance (SList.default acc) k rest
+    | true :: rest -> remake_instance (SList.cons (mkRel k) acc) (k-1) rest
+  in
+  let rec aux (k,bnd) env env' =
+    match env, env' with
+    | [], [] -> mkEvar (ev, remake_instance SList.empty k (List.rev bnd))
+    | d :: env, d' :: env' ->
+      (* is_section_variable: inserting letins for section variables
+         can produce spurious dependencies on them *)
```

Comment:
I don't know what you mean.

Reply:
In more detail:
I see two categories of errors on CI on this PR:
1. Qed-errors that are all of the form "Error: In environment ... The term "..." has type "..." while it is expected to have type "..."." (fiat-parsers, flocq, color, coqprime, iris, math_classes, mczify, neural_net_interp, unimath, atbr, perennial, relation_algebra, rewriter, smtcoq, fiat_crypto_legacy).  ("Error: Recursive definition of cf is ill-formed." in verdi_raft probably also has the same root cause.)  I am guessing that there's a de Bruijn indexing bug, because these all look like errors where it's getting a proof as an argument where it's supposed to be getting the object the proof is about.  (If it's not the offsets, maybe it's the `List.rev`?  But I'm just guessing here.)
2. I also see errors such as
  - "Error: This proof is focused, but cannot be unfocused this way" (HoTT), "Error: e is used in hypothesis g." (metacoq),
  - "Error: No applicable tactic." (bedrock2),
  - coq_dpdgraph has what looks like a change in dependencies between terms,
  - "Non-exhaustive pattern-matching, no clause found for: dummy ?(nth i Γ0 unit) ?(i) f arg Γ' f' A' B' prf t o Heqhsubst H t0 o0 Heqhsubst0 H0 H1 Heq Hind ?(Γ0) T" and "Error: Could not find an instance for "FunctionalElimination (@comp_assoc_simpl_rec)  ?fun_elim_ty ?n"." in equations_test, 

I am hypothesizing that the second kind of errors might be a result of changing which section variables some definitions depend on (by changing what let-ins they depend on) which might change type signatures.  (coq_dpdgraph actually seems like the strongest evidence that something like this might be happening)

(tactician is just a missing an overlay adding a case for REVERTcast to OCaml code)

Reply:
Oh, I see, you're saying that we avoid inserting let-ins for section variables, so section variables should not change dependencies.